### PR TITLE
[dashboard] Update Project Configurator

### PR DESCRIPTION
**Pull request by jankeromnes**
Addresses part of https://github.com/gitpod-io/gitpod/issues/4948 (just the final Configuration part)

### Specs

https://github.com/gitpod-io/gitpod/issues/4948#issuecomment-895927769

### Details

- Imports https://github.com/gitpod-io/gitpod-yml-inferrer and uses it to try and guess a project's setup
- If there is a repo-based config, that always takes priority
- If there none, it tries to guess the setup (see all supported `checkX` setups in [config-inferrer.ts](https://github.com/gitpod-io/gitpod-yml-inferrer/blob/main/src/config-inferrer.ts))
- If it doesn't guess the setup, it uses a default config (which cannot be parsed, due to comments, and to force users to enter something -- but maybe this isn't great)
- You can run a prebuild based on the config, and see the log output & status (the "New Workspace" button doesn't work yet 🙈)

### How to test

- Use Admin to enable the Teams & Projects feature flag
- Add a project with a repo-based setup (any repo with a `.gitpod.yml`) -- repo-based always has priority
- Add a project with a recognizable setup (e.g. any repo with a `package.json`)
- Add a project with an unrecognizable setup (e.g. an empty repo)

For each of these, try the Project Configurator.

- E.g., add empty repo, change the config to `init: yarn install` and `command: yarn start`, click `Run Prebuild`

### Screenshots

<table><tr><td>
Screenshot 2021-08-19 at 17 54 59
</td><td>
Screenshot 2021-08-19 at 17 53 25
</td><td>
Screenshot 2021-08-19 at 17 53 38
</td><td>
Screenshot 2021-08-19 at 17 53 55
</td><td>
Screenshot 2021-08-19 at 17 54 27
</td></tr></table>

/uncc

- [x] /werft with-clean-slate-deployment

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **jankeromnes** after the PR has been reviewed.

*No associated issue*. Update pull-request body to add a reference to an issue, or get approval with `/approve no-issue`

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**
- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

Approval requirements bypassed by manually added approval.

This pull-request has been approved by:

*No associated issue*. Requirement bypassed by manually added approval.

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**
- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by AlexTugarev** [comment]
Nice just tried, and really like the new look!

About the `Project type detected` hint/message, it appear to be quite dominant like the other warnings. I think it would work without it as well. Also because it repeats the instructions on the right. 

Screen Shot 2021-08-20 at 11 11 56

**Comment by gtsiolis** [comment]
Looking a this now! 👀

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/ConfigureProject.tsx:154]: **suggestion**: @svenefftinge brought up an idea to allow users to resize the editor width, see [relevant discussion](https://gitpod.slack.com/archives/C021Q4BL8E8/p1629463581276700?thread_ts=1629461726.275900&cid=C021Q4BL8E8) (internal). This would require to have a visual indicator for users to grab and move around. For the first iteration we could go with a 50/50 split which also seems nice. What do you think @jankeromnes? 

| BEFORE | AFTER |
|-|-|
| Screenshot 2021-08-20 at 4 27 44 PM | Screenshot 2021-08-20 at 4 27 38 PM |

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/ConfigureProject.tsx:210]: **suggestion**: Do you think adding some trailing dots are not helpful?

```suggestion
        <span>Prebuild in progress ...</span>
```

Comment (gtsiolis) [inline components/dashboard/src/projects/ConfigureProject.tsx:210]: **praise**: This one suddenly moved to the top of my favorite visual elements within the product.

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/ConfigureProject.tsx:176]: **question**: Do you think it's ok to leave this message here when prebuilds are being triggerred from push events or the branches pages?

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/ConfigureProject.tsx:221]: **issue**: Prebuilds run in the configuration page do not surface in the prebuilds page. I'd think all triggered prebuilds here would end up in the prebulds page. What do you think? Cc @AlexTugarev @svenefftinge

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/ConfigureProject.tsx:232]: **issue**: When prebuild finishes, log output disappears.

Comment (gtsiolis) [inline components/dashboard/src/projects/ConfigureProject.tsx:232]: **issue**: Reloading when a prebuild finishes, causes the log output to disappear.

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/ConfigureProject.tsx:184]: **issue**: Clicking run prebuild when a prebuild is running, causes the log output to disappear.

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/ConfigureProject.tsx:168]: **issue**: When adding an empty repository, this is stuck into detecting loop. Could we use an error message and prompt the user to open a new workspace instead?

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/ConfigureProject.tsx:184]: **suggestion**: We could disable the run prebuild button when a prebuild is in progress. Alternatively, to avoid disabling buttons, which is a bad practice in terms of accessibility, we could trigger a modal to confirm a new prebuild and cancel the previous one in the background.

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/ConfigureProject.tsx:168]: **issue**: I could not get a detectable project up and running. 😭

File: components/dashboard/src/components/MonacoEditor.tsx
Comment (gtsiolis) [inline components/dashboard/src/components/MonacoEditor.tsx:18]: **praise**: Thanks for adding dark theme support for the editor! 🌕

File: components/dashboard/src/components/WorkspaceLogs.tsx
Comment (gtsiolis) [inline components/dashboard/src/components/WorkspaceLogs.tsx:15]: **praise**: Thanks for using the same color scale! 🖍️

**Review by gtsiolis — COMMENTED** [review]
C'est magnifique @jankeromnes! 🔮 

Thanks for pushing this forward and adding support for dark theme layouts! 🙇 

Left a first round of comments. If you think any of these comments significantly increase the scope of this PR let's open follow up issues if needed.

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (jankeromnes) [reply components/dashboard/src/projects/ConfigureProject.tsx:184]: Yes (it doesn't work as expected yet -- will open a follow-up issue to fix & enable it)

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (jankeromnes) [reply components/dashboard/src/projects/ConfigureProject.tsx:183]: Also yes 😬

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

Approval requirements bypassed by manually added approval.

This pull-request has been approved by:

Associated issue: *#4948*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**
- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (jankeromnes) [reply components/dashboard/src/projects/ConfigureProject.tsx:162]: @gtsiolis Sure! But which gray? The [specs](https://www.figma.com/file/iYeLdIMjMdSpTgFGEAicmk/Project-Configuration-%234648?node-id=0%3A1) have a white background in every non-disabled editor.

Comment (gtsiolis) [reply components/dashboard/src/projects/ConfigureProject.tsx:162]: > The specs have a white background in every non-disabled editor.

That's correct!

> But which gray?

@jankeromnes What do you think of using `gray-400` for the text and `gray-100` for the background when the editor is disabled? See [relevant frame in specs](https://www.figma.com/file/iYeLdIMjMdSpTgFGEAicmk/Project-Configuration-4648?node-id=1417%3A1781).

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (jankeromnes) [reply components/dashboard/src/projects/ConfigureProject.tsx:168]: > When adding an empty repository, this is stuck into detecting loop.

Oh, that's unexpected & sounds like a bug. For me, an empty repository gets "detected" as "no setup found", and immediately gets the default config.

Did you notice any errors in the browser console?

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (jankeromnes) [reply components/dashboard/src/projects/ConfigureProject.tsx:162]: > That's correct!

Aha, "when the configuration is non-editable" was common to both "a)" and "b)" 😅 (sorry, I understood "a) (always) and b) (when disabled)")

> What do you think of using `gray-400` for the text and `gray-100` for the background when the editor is disabled?

Sure! But won't that be problematic in terms of minimum acceptable contrast? (I know for me it might cause some eye strain.) Happy to try the suggested colors as a first iteration!

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (jankeromnes) [reply components/dashboard/src/projects/ConfigureProject.tsx:154]: Could work! I'll give this a try, thanks.

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (jankeromnes) [reply components/dashboard/src/projects/ConfigureProject.tsx:210]: The specs don't have dots. I could try, but feel a bit awkward about them (because there is normally no whitespace before an ellipsis, and also there is a unicode character `…` that should be used instead of the three dots `...` 😅 also in this case, we also have a spinner that sort of duplicates the trailing dots' meaning -- but, no strong opinion here actually)

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (gtsiolis) [reply components/dashboard/src/projects/ConfigureProject.tsx:162]: > But won't that be problematic in terms of minimum acceptable contrast?

You're right! Thanks for caring about accessibility. What do you think of chaning the background to `gray-100` but keeping the text color as is (`gray-600`) as this combination passes `WCAG AA` color checks?

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (gtsiolis) [reply components/dashboard/src/projects/ConfigureProject.tsx:210]: > The specs don't have dots.

I'd swear I saw them somewhere but maybe in an earlier design! Let's skip adding them for now unless you think these are helpful. 🙀

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (jankeromnes) [reply components/dashboard/src/projects/ConfigureProject.tsx:221]: I think they should, and it could be a bug that they don't (the configurator currently saves the project configuration and then triggers a regular prebuild for the default branch)

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (jankeromnes) [reply components/dashboard/src/projects/ConfigureProject.tsx:232]: Strange -- for me, the logs never show up when the prebuild is running, and only appear after it has finished. 🤔 Sounds like we need to do some logs debugging.

Comment (jankeromnes) [reply components/dashboard/src/projects/ConfigureProject.tsx:232]: Ah, true, we're not actually fetching prebuilds here 😬 maybe we should display the latest successful prebuild here? (And also maybe check if the configuration was the same)

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (jankeromnes) [reply components/dashboard/src/projects/ConfigureProject.tsx:168]: That's unexpected. @AlexTugarev seems to have gotten one in https://github.com/gitpod-io/gitpod/pull/5236#issuecomment-902558545 -- which project did you try?

Hint: Any project with a `package.json` and _without_ a `.gitpod.yml` should get successfully detected (even if private).

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (jankeromnes) [reply components/dashboard/src/projects/ConfigureProject.tsx]: I'd feel a bit strange with the proposed new copy:

> Prebuilds run on every commit

As a user discovering the Project Configurator, I'm now faced with a conflict: Should I click the "Run Prebuild" button, or push new commits instead as suggested here?

I don't think commits are relevant here, and even add confusion (either you're in a workspace, pushing commits, or you're in the Project Configurator and you test a config with a few clicks).

Comment (gtsiolis) [reply components/dashboard/src/projects/ConfigureProject.tsx]: You are probably right here! The intent here was to _refrain_ from a) having duplicate content as the auto-detection message on the left[[1](https://github.com/gitpod-io/gitpod/pull/5236/files#diff-f4b8adb02a103d42c0de94bec0d32a6179440371ee01c21336a2181fcfa3a66dR110)][[2](https://github.com/gitpod-io/gitpod/pull/5236/files#diff-f4b8adb02a103d42c0de94bec0d32a6179440371ee01c21336a2181fcfa3a66dR118)] and b) showing an inaccurate message as editing the configuration is not possible for projects with a git-based configuration already in place. Instead this could help educate users about prebuilds.

What do you think of:

1. Keeping the following on the empty state on the right:

> Run prebuild or open a new workspace to edit project configuration.

2. Changing the message on the left when the project has been detected successfully to something like the following:

> The starter template below has been added based on the project type detected.

🍊 🍊 🍊 🍊 

Also, enabling the _Run Prebuild_ button would be great here, right?

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (jankeromnes) [reply components/dashboard/src/projects/ConfigureProject.tsx]: Sounds good to me, thanks! Left the text color as is, but adjusted the background according to whether editing is disabled or not (in both light and dark themes). Result looks great 😁 ✨

File: components/dashboard/src/projects/ConfigureProject.tsx
Comment (gtsiolis) [reply components/dashboard/src/projects/ConfigureProject.tsx:168]: > Did you notice any errors in the browser console?

Do not recall but will try again once the new preview env is up and running.

Comment (gtsiolis) [reply components/dashboard/src/projects/ConfigureProject.tsx:168]: > ... which project did you try?

I remember trying a private repository with a starter template with Next.js.

**Comment by jankeromnes** [comment]
Okay, this is about as far as I'll get in this PR.

@AlexTugarev could you please give this another look, and approve if it can be merged? I can adjust minor things and migrate all remaining open discussions to follow-up issues.

**Heads up**: This refactors `WorkspaceLogs`, i.e. the logs component used everywhere in the dashboard (even in non-Teams & Projects pages, like the workspace start page). I've tried testing that logs didn't break, but there are too many random logs problems* on dev-staging currently so this can't really be tested. Hopefully this can be test on staging before it goes live.

*Problems that make testing the logs almost impossible:
<table><tr><td>Screenshot 2021-08-24 at 14 58 42</td>
<td>Screenshot 2021-08-24 at 15 47 48</td>
<td>Screenshot 2021-08-24 at 15 55 59</td></tr></table>

**Review by svenefftinge — APPROVED** [review]

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: db3b657b23d0eb673221714f80590cfbd6eaddc5</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/5236#pullrequestreview-737932905" title="Approved">svenefftinge</a>*

Associated issue: *#4948*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [svenefftinge]
- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [svenefftinge]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [svenefftinge]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [svenefftinge]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
